contributors/design-proposals/storage/csi-migration.md

```diff
@@ -661,6 +721,111 @@ with minimal disruption to users would be:
 And at this point users doing their own deployment and not installing the GCE PD
 CSI driver encounter an error.
 
+## Disabling in-tree plugin code
+
+Before we can stop compiling and prepare to remove the code associated with migrated
+in-tree plugins, we need to make sure all persistent volume operations involving
+in-tree plugins continue to function in a backward compatible way when the
+in-tree plugin code paths are disabled. When CSIMigrationInTreeOff feature flag
+is enabled, we will not invoke ProbeVolumePlugins() for the in-tree plugins (that
+have plugin-specific migration feature flag enabled) in appendAttachableLegacyProviderVolumes()
+and appendLegacyProviderVolumes() in the Kubernetes Controller Manager and Kubelet.
+All functions in Kubernetes code base that depend on probed in-tree plugins need to be
+audited and refactored to handle errors returned (due to absence of a probed plugin)
+as detailed below.
+
+### Enhancements in Probing/Registration of in-tree plugins
+
+Functions appendLegacyProviderVolumes (in the Kubernetes Controller Manager and Kubelet)
+and appendAttachableLegacyProviderVolumes (in the Kubernetes Controller Manager)
+will be enhanced to [1] check plugin specific migration feature flags as well as
+[2] the overall CSIMigrationInTreeOff feature flag to determine whether ProbeVolumePlugins
+function of a legacy in-tree plugin will get invoked.
+
+Once CSIMigrationInTreeOff feature flag and the plugin specific migration
+flags get enabled by default, the build tag `nolegacyproviders` can be enabled on
+that no in-tree plugins get probed.
+
+### Detection of migration status for a plugin
+
+Code paths that need to check migration status of a plugin, for example,
+in provisionClaimOperationExternal, findDeletablePlugin, etc. will need to
+depend on an instance of the CSIMigratedPluginManager that provides the following
+utilities:
+```
+func (pm CSIMigratedPluginManager) IsCSIMigrationEnabledForPluginByName(pluginName string) bool
+func (pm CSIMigratedPluginManager) IsPluginMigratableToCSIBySpec(spec *Spec) (bool, error)
+```
+The CSIMigratedPluginManager will be introduced in pkg/volume/csi_migration.go
+
+Note that a per-plugin member function of the form IsMigratedToCSI cannot be used
+since [1] a plugin object for in-tree plugins will typically be nil and [2] the
+code implementing IsMigratedToCSI will be removed as part of disabling plugin code.
+
+### Handling of errors returned by FindPluginBySpec/FindPluginByName for legacy plugins
+
+Once an in-tree plugin is no longer probed (through ProbeVolumePlugins), all the VolumePluginMgr
+functions of the form Find*PluginBySpec/Find*PluginByName will return an error.
+For example, invocation of:
+
+1. FindProvisionablePluginByName in the pv controller will return error for a migrated
```

Comment:
IIUC these functions should never actually get invoked with a `Spec` or `Name` of an in-tree plugin anymore right? They should have been migrated before these `Find` functions are invoked. If an error does occur that's actually because of a real error in either the migration logic or something else right?

Reply:
Not when we specifically use the error code to branch out like this: https://github.com/kubernetes/kubernetes/pull/83098/files#diff-3f32ad90966ceeb297f57c2a8d726348R1673 . Otherwise, yes it's a logic error.

I point out ```Code invoking the above functions will need to check for migration status of a plugin
based on Volume spec or Plugin name rather than throwing an error.``` below for code like https://github.com/kubernetes/kubernetes/pull/83098/files#diff-6db487adca84bc74a8dc0538c8dd6463R256